### PR TITLE
Custom SDP implementation runtime dispatch

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This shows the run mechanism for dispatching to a custom backend kernel or defaulting to the composite version for SDP. Currently we short circuit for composite but in the future this will dispatch to the appropriate kernel.

### CODE REVIEW

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 1]
```diff
@@ -217,36 +211,142 @@ Tensor NestedTensor_to_padded_tensor_cuda(
...
 }
 
-Tensor flash_scaled_dot_product_attention(
+std::tuple<Tensor, Tensor> _scaled_dot_product_attention_forward_nested(
+        const Tensor& query_, const Tensor& key, const Tensor& value,
+        const c10::optional<Tensor>& attn_mask_, double dropout_p, bool need_attn_weights, bool is_causal) {
+
+    // Determine which efficient kernel to use
+    sdp::sdp_params kernel_params{query_, key, value, attn_mask_.has_value(), dropout_p, is_causal};
+    if (sdp::use_flash_attention(kernel_params)){
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:220]: replace with  select_sdp_backend

**Comment by drisspg** [comment]
Notes:
- Did not implement context manager versions of use_fused and use_math
- Not tested (YAY!). Not totally true since we have test_transformers that exercised fused_sdp function. However we do not build with flash_attention on CI/CD. As well we don't have A100 that can execute the kernels. SO not totally sure how to do this.

File: aten/src/ATen/native/transformers/sdp_utils.h [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+      at::kHalf, at::kBFloat16};
+
+  //
+  std::vector<std::function<bool(sdp_params, bool)>> constraints{
+      check_runtime_disabled,
+      check_tensor_shapes,
+      check_packed_projection_shape,
+      check_for_attn_mask,
+      check_cuda_tensors,
+      compiled_with_fused_kernels};
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/sdp_utils.h:214]: Add a cuda arch check?

Comment (drisspg) [reply aten/src/ATen/native/transformers/sdp_utils.h]: added. Lets see if CI likes this change

File: aten/src/ATen/Context.h [hunk 3]
```diff
@@ -126,6 +126,19 @@ class TORCH_API Context {
   bool deterministicCuDNN() const;
   void setDeterministicCuDNN(bool);
 
+  // Note [Disabling Fused SDP Kernels]
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // Fused SDP kernels are enabled by default. However, they can be disabled
+  // by setting the at::globalContext().setDisableFusedSDP(true) flag.
```
Comment (cpuhrsch) [inline aten/src/ATen/Context.h:132]: nit: did you mean to update `setDisableFusedSDP`?

File: aten/src/ATen/Context.h [hunk 4]
```diff
@@ -126,6 +126,19 @@ class TORCH_API Context {
...
+  // This is useful for debugging purposes. For example, if you want to
+  // compare the performance of the fused SDP kernels with the unfused
+  // kernels, you can disable the fused SDP kernels.
```
Comment (cpuhrsch) [inline aten/src/ATen/Context.h:135]: You can also force your code to use fused  kernels only if you disable the math kernels.

File: torch/backends/cuda/__init__.py [hunk 5]
```diff
@@ -160,5 +161,21 @@ def preferred_linalg_library(backend: Union[None, str, torch._C._LinalgBackend]
 
     return torch._C._get_linalg_preferred_backend()
 
+def fused_sdp_enabled():
+    r""" Returns whether fused sdp is enabled or not. """
+    return torch._C._get_fused_sdp_enabled()
+
+def set_fused_sdp_enabled(enabled: bool):
+    r""" Enables or disables fused sdp. """
+    torch._C._set_enabled_fused_sdp(enabled)
```
Comment (cpuhrsch) [inline torch/backends/cuda/__init__.py:170]: nit: there's two verbs and each in a different tense. Why not `set_sdp_use_fused` or `enable_fused_sdp`?

File: aten/src/ATen/Context.h (hunk 4 above)
Comment (drisspg) [reply aten/src/ATen/Context.h:135]: will add

File: torch/csrc/Module.cpp [hunk 6]
```diff
@@ -467,7 +467,36 @@ PyObject* THPModule_float32MatmulPrecision(
   }
   return THPUtils_packString(s);
 }
-
+PyObject* THPModule_setUserEnabledFusedSDP(PyObject* _unused, PyObject* arg) {
+  THPUtils_assert(
+      PyBool_Check(arg),
+      "set_enabled_fused_sdp expects a bool, "
```
Comment (drisspg) [inline torch/csrc/Module.cpp:473]: change this message

Comment (drisspg) [reply torch/csrc/Module.cpp:473]: and the ones below to match the function names

**Pull request closed without merging**